crates/holochain_zome_types/src/signature.rs

```diff
@@ -79,3 +79,47 @@ impl SignEphemeral {
         self.0
     }
 }
+
+/// Some data with a signature attached
+#[derive(
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+    Serialize,
+    Deserialize,
+    derive_more::Constructor,
+    derive_more::Deref,
+    derive_more::From,
+    derive_more::Into,
+)]
+#[cfg_attr(feature = "fuzzing", derive(arbitrary::Arbitrary))]
+pub struct Signed<T>
+where
+    T: serde::Serialize + serde::de::DeserializeOwned,
+{
+    #[deref]
+    #[serde(bound(deserialize = "T: serde::de::DeserializeOwned"))]
+    data: T,
+    signature: Signature,
+}
```

Comment:
I'd just like to point out data structures like this perpetuate the incorrect practice of pairing signatures with unserialized data. We sign the serialized data. When we verify signatures, we verify them against the serialized data. We wouldn't have any problems with non-deterministic serialization if we paired these properly.

Reply:
https://github.com/holochain/holochain/blob/414e21ec9ad00023fab160721f9e1e9616e4ebe1/crates/kitsune_p2p/types/src/agent_info.rs#L97-L101

Reply:
Yes good point, let me find out how to rework this

Reply:
I see what you're saying, and this is a deeper problem of representation. Seems we'd have to change a bunch of things in a lot of places to get this right. I introduced this because I needed the parallel pattern of what Actions are doing, for Warrants. Maybe we can reframe this more desirably by noticing that this new type generalizes a bad pattern, so that if we ever improve on the pattern like you say, this type will be a pivot point for that refactor?

I'll also add this context to the doc